### DISCUSSION TITLE
How to preprocess a single image for inference

### DISCUSSION THREAD

**Discussion by SarBH** [qubvel/segmentation_models.pytorch#444] [Q&A] [answered]
Is there a simple way to prepare a single image to go through a trained model and output the mask?

Context:
I went through your CamVid car segmentation example. Now I have a trained model.
If I wanted to simply load the model to run inference on a single example without having to create a `torch.utils.data.Dataset and Dataloader`, whats the best way to do this? 

I was hoping to do:
```
preprocessing_fn = smp.encoders.get_preprocessing_fn(args.encoder, args.encoder_weights)
processed = preprocessing_fn(img)

```

**Comment by SarBH** [comment, marked as answer]
UPDATE: if we want the preprocessing function to behave as i described `processed = preprocessing_fn(img)` you can use
```
    image = cv2.imread(args.sample_input)
    comp = get_preprocessing(preprocessing_fn)
    x = comp(image=image)
```
```
def get_preprocessing(preprocessing_fn):
    """Construct preprocessing transform
    
    Args:
        preprocessing_fn (callbale): data normalization function 
            (can be specific for each pretrained neural network)
    Return:
        transform: albumentations.Compose
    
    """
    
    _transform = [
        albu.Lambda(image=preprocessing_fn),
        albu.Lambda(image=to_tensor, mask=to_tensor),
    ]
    return albu.Compose(_transform)

```

**Reply by YoniChechik** [reply]
I believe this is incorrect. totensor also normalizes the image (see: https://pytorch.org/vision/stable/generated/torchvision.transforms.ToTensor.html#torchvision.transforms.ToTensor), which you dont want to do because the preprocessing_fn already does this.

**Reply by tatyana-pichugina** [reply]
Be careful,cv2 trap. cv2.imread(args.sample_input) by default reads BGR instead RGB.
